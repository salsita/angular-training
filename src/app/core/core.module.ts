import { HttpClientModule } from '@angular/common/http';
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { ApiModule } from '@salsita/ng-api';
import { CrudModule } from '@salsita/ng-crud';
import { EntityRepositoryModule } from '@salsita/ng-entities';
import { RouterModule as RouterStoreModule } from '@salsita/ng-router';

import { environment } from '../../environments/environment';
import { SharedModule } from '../shared/shared.module';
import { ApiErrorToastComponent } from './api/api-error-toast/api-error-toast.component';
import { ApiLoaderComponent } from './api/api-loader/api-loader.component';

@NgModule({
  imports: [
    HttpClientModule,

    ApiModule.forRoot({ storeNamespace: 'api', baseUrl: environment.apiBaseUrl }),
    CrudModule.forRoot({ storeNamespace: 'crud' }),
    EntityRepositoryModule.forRoot({ storeNamespace: 'entityRepository' }),
    RouterStoreModule.forRoot({ storeNamespace: 'router' }),
    SharedModule
  ],
  declarations: [ApiErrorToastComponent, ApiLoaderComponent],
  exports: [ApiErrorToastComponent, ApiLoaderComponent, ApiModule, HttpClientModule]
})
export class CoreModule {
  constructor(
    @Optional()
    @SkipSelf()
    parentModule: CoreModule
  ) {
    if (parentModule) {
      throw new Error(
        `CoreModule has already been loaded. Import Core modules in the AppModule only.`
      );
    }
  }
}
