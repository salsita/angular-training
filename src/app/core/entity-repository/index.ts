export {
  EntityRepositoryModuleConfig,
  EntityRepositoryState,
  EntityRepositoryStateRoot,
  SingleEntityRepository
} from './entity-repository.interfaces';
export { EntityRepositorySelectors } from './entity-repository.selectors';
export { EntityRepositoryService } from './entity-repository.service';
export { EntityRepositoryModule } from './entity-repository.module';
