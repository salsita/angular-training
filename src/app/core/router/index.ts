export {
  FlatRouterStateSnapshot,
  RouterModuleConfig,
  RouterState,
  RouterStateRoot
} from './router.interfaces';
export { RouterSelectors } from './router.selectors';
export { CustomRouterStateSerializer } from './router.serializer';
export { RouterModule } from './router.module';
