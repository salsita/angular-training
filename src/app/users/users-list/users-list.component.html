<div>
  <table>
    <thead>
      <tr>
        <th>First Name</th>
        <th>Last Name</th>
      </tr>
    </thead>
    <tbody>
      <tr *ngFor="let user of (users$ | async)">
        <td>
          <a [routerLink]="['/users', user.id]">{{user.firstName}}</a>
        </td>
        <td>{{user.lastName}}</td>
      </tr>
    </tbody>
  </table>
  <a [routerLink]="['/users/create']">Add user</a>
</div>
<br>

<router-outlet></router-outlet>
