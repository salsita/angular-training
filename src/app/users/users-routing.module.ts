import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import { SkillsResolver } from './skills/skills.resolver';
import { UserCreateComponent } from './user-create/user-create.component';
import { UserDetailComponent } from './user-detail/user-detail.component';
import { UserDetailResolver } from './user-detail/user-detail.resolver';
import { UsersListComponent } from './users-list/users-list.component';
import { UsersListResolver } from './users-list/users-list.resolver';


const routes: Routes = [
  {
    path: '',
    resolve: {
      skills: SkillsResolver
    },
    children: [
      {
        path: '',
        component: UsersListComponent,
        resolve: {
          users: UsersListResolver
        },
        children: [
          {
            path: 'create',
            component: UserCreateComponent
          }
        ]
      },
      {
        path: ':userId',
        component: UserDetailComponent,
        resolve: {
          user: UserDetailResolver
        }
      }
    ]
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
  providers: []
})
export class UsersRoutingModule { }
