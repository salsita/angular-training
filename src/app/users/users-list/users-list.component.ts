import { Component } from '@angular/core';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';

import { getUsersList } from '../+state/users.selectors';
import { AppState } from '../../app.interfaces';
import { User } from '../users.interfaces';

@Component({
  selector: 'app-users-list',
  templateUrl: './users-list.component.html',
  styleUrls: ['./users-list.component.scss']
})
export class UsersListComponent {
  users$: Observable<User[]>;

  constructor(store: Store<AppState>) {
    this.users$ = getUsersList(store);
  }
}
