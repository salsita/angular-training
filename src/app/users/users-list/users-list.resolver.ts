import { Injectable } from '@angular/core';

import { userSchema } from '../+state/users.schema';
import { withLoadingIndicator } from '../../core/api';
import { CrudResolver, CrudService } from '../../core/crud';
import { RouterSelectors } from '../../core/router';
import { UsersApi } from '../users.api';

@Injectable()
export class UsersListResolver extends CrudResolver {
  schema = [userSchema];
  route = 'users/list';
  key = 'users';

  constructor(private api: UsersApi, crud: CrudService, routerSelectors: RouterSelectors) {
    super(crud, routerSelectors);
  }

  @withLoadingIndicator()
  data() {
    return this.api.getUsers();
  }
}
