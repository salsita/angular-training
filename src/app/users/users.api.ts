import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { apiCall } from '../core/api/api.decorators';
import { CrudId } from '../core/crud/crud.interfaces';
import { User } from './users.interfaces';

@Injectable()
export class UsersApi {
  constructor(private http: HttpClient) {}

  @apiCall()
  getUsers() {
    return this.http.get(`/users`);
  }

  @apiCall()
  getUser(id: CrudId) {
    return this.http.get(`/users/${id}`);
  }

  @apiCall()
  deleteUser(id: CrudId) {
    return this.http.delete(`/users/${id}`);
  }

  @apiCall()
  saveUser({ id, ...data }: User) {
    return id ? this.http.patch(`/users/${id}`, data) : this.http.post('/users', data);
  }

  @apiCall()
  getSkills() {
    return this.http.get(`/skills`);
  }
}
