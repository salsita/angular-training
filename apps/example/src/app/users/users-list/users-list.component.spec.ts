import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed, async } from '@angular/core/testing';

import { UsersSelectors } from '../+state/users.selectors';
import { MockUsersSelectors } from '../testing';
import { UsersListComponent } from './users-list.component';

describe('UsersListComponent', () => {
  let fixture: ComponentFixture<UsersListComponent>;

  beforeEach(
    async(() => {
      TestBed.configureTestingModule({
        declarations: [UsersListComponent],
        schemas: [NO_ERRORS_SCHEMA],
        providers: [{ provide: UsersSelectors, useClass: MockUsersSelectors }]
      }).compileComponents();
    })
  );

  beforeEach(() => {
    fixture = TestBed.createComponent(UsersListComponent);
    fixture.detectChanges();
  });

  it('render list of users', () => {
    expect(fixture).toMatchSnapshot();
  });
});
