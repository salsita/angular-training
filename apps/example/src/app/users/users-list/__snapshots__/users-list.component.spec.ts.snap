// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersListComponent render list of users 1`] = `
<app-users-list
  users$={[Function Observable]}
>
  <div>
    
    
    <table>
      
      
      <thead>
        
        
        <tr>
          
          
          <th>
            First Name
          </th>
          
          
          <th>
            Last Name
          </th>
          
        
        </tr>
        
      
      </thead>
      
      
      <tbody>
        
        
        
        <tr>
          
          
          <td>
            
            
            <a>
              John
            </a>
            
          
          </td>
          
          
          <td>
            Doe
          </td>
          
        
        </tr>
        <tr>
          
          
          <td>
            
            
            <a>
              John
            </a>
            
          
          </td>
          
          
          <td>
            Stiles
          </td>
          
        
        </tr>
        
      
      </tbody>
      
    
    </table>
    
    
    <a>
      Add user
    </a>
    
  
  </div>
  <br />
  
  <router-outlet />
  
</app-users-list>
`;
